### ISSUE TITLE
Validation step is ignored when using DataModule

### ISSUE THREAD

**Issue by ncuxomun** [PyTorchLightning/pytorch-lightning#5454]
#### What is your question?
Hi, guys!
I created my own DataModule and loading it to the trainer. However, it appears that the "fit" is skipping the validation step.
How can I ensure that the code runs through the validation step too?

#### Code
```
class DataModule(pl.LightningDataModule):
    def __init__(self, batch_size=25, seed=0):
    # def __init__(self, dataset, batch_size=25, seed=0):
        super().__init__()
        self.dataset = dataset
        self.batch_size = batch_size
        self.seed = seed
        self.split = [801, 100, 100]
        # self.transform = torchvision.transforms.ToTensor()

    def setup(self, stage=None):
        # train/valid/test split
        # and assign to use in dataloaders via self
        train_set, valid_set, test_set = torch.utils.data.random_split(self.dataset, self.split, generator=torch.Generator().manual_seed(self.seed))

        if stage == 'fit' or stage is None:

            self.train_set = train_set
            self.valid_set = valid_set

        if stage == 'test' or stage is None:
            self.test_set = test_set

    def train_dataloader(self):
        return torch.utils.data.DataLoader(self.train_set, batch_size=self.batch_size, shuffle=True)

    def valid_dataloader(self):
        return torch.utils.data.DataLoader(self.valid_set, batch_size=self.batch_size, shuffle=False)

    def test_dataloader(self):
        return torch.utils.data.DataLoader(self.test_set, batch_size=self.batch_size, shuffle=False)

class LitReg(pl.LightningModule):
    def __init__(self, in_dims, out_dims, lr=2e-4, max_dict={}):
        super().__init__()
        self.in_size = in_dims
        self.out_size = out_dims
        self.lr = lr
        self.max_dict = max_dict

        # model
        self.model = nn.Sequential(
            nn.Linear(self.in_size, self.in_size),

            nn.LeakyReLU(0.02),

            nn.Linear(self.in_size, self.out_size)
        )

        self.model.apply(self.weights_init)

    def forward(self, data):
        out = self.model(data)
        return out

    def training_step(self, batch, batch_idx):
        x, y, l_rate = batch
        pred_y = self.model(x)
        train_loss = F.mse_loss(pred_y, y) 

        self.log('train_loss', train_loss, prog_bar=True)
        return train_loss

    def validation_step(self, batch, batch_idx):
        self._shared_eval(batch, batch_idx, 'val')

    def test_step(self, batch, batch_idx):
        self._shared_eval(batch, batch_idx, 'test')

    def _shared_eval(self, batch, batch_idx, prefix):
        x, y, l_rate = batch
        pred_y = self.model(x)

        loss = F.mse_loss(pred_y, y) 
        self.log(f'{prefix}_loss', loss, prog_bar=True)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), self.lr)
        return optimizer

    def weights_init(self, m):
        if isinstance(m, nn.Conv2d) or isinstance(m, nn.ConvTranspose2d):
            torch.nn.init.normal_(m.weight, 0.0, 0.02)
        if isinstance(m, nn.BatchNorm2d) or isinstance(m, nn.BatchNorm1d):
            torch.nn.init.normal_(m.weight, 0.0, 0.02)
            torch.nn.init.constant_(m.bias, 0)
        if isinstance(m, nn.Linear):
            torch.nn.init.normal_(m.weight, 0.0, 0.02)
            torch.nn.init.constant_(m.bias, 0)
```
#### What have you tried?
Placing breakpoints to debug in VSCode, but all in vain.
Also accessed both valid and test datasets and loaders. All looks set.
What is working? If I load the data the following way.
```
    **train_loader = DataLoader(X_train, batch_size=args.batch_size)
    val_loader = DataLoader(X_val, batch_size=args.batch_size)
    test_loader = DataLoader(X_test, batch_size=args.batch_size)

    trainer.fit(model, train_loader, val_loader)**
```

#### What's your environment?

 - OS: Win
 - Packaging pip
 - Version 1.1.3

Thank you for your attention!

**Comment by s-rog** [comment]
```
    def validation_step(self, batch, batch_idx):
        return self._shared_eval(batch, batch_idx, 'val')
```
You forgot return!

**Comment by ncuxomun** [comment]
> ```
>     def validation_step(self, batch, batch_idx):
>         return self._shared_eval(batch, batch_idx, 'val')
> ```
> 
> You forgot return!

Thanks for the remark! I corrected the mistake. But I still get the same output, i.e. the validation step is skipped. Is there anything else that I am missing?

**Comment by s-rog** [comment]
@ncuxomun 
You're not logging anything in validation_step and you have not implemented validation_epoch_end (no logging there as well) so there would be no log output

**Comment by s-rog** [comment]
Oh I see that you called self.log in _shared_eval, so the return shouldn't be necessary, apologies.

Does the script run with the validation progress bar (at the end of a training epoch)?

**Comment by ncuxomun** [comment]
Thanks for re-opening it. I was looking your your earlier comment))
Yes, it is connected via the shared function here. I also tried without it, that is enabling both val and test function, and disabling the the shared one.
No, the script does not not run with the validation progress bar. It displays "loss" and "train_loss" only. Another observation is that what I did include "return" and "fit by trainer.fit(model, datamodule=dm), the training took more epochs, ignoring the early stopping callback (which is honored when I don't work with datamodule).

**Comment by s-rog** [comment]
Could you try initializing the dataset in datamodules setup instead of passing it in as an arg?

**Comment by ncuxomun** [comment]
Cool, thanks for suggestions. Here's what I did:
1. Dataset was moved into prepare_data, where it was initialized.
2. Then I placed a breakpoint into the dataset to check its dimensions. This prompted the following warning, which I did not see before: 
_"UserWarning: you defined a validation_step but have no val_dataloader. Skipping validation loop warnings.warn(*args, *kwargs)_"
3. Then, I renamed "_valid_dataloader_" into "**val_dataloader**", re-run then case and all progress bar values were displayed as required. 
It was my bad, I should have called that function precisely as the said in the docs. One thing that is surprising is that this "skipping" part was not a big deal when I was loading the train_dataloader and val_dataloader individually instead of datamodule. Do you know what could be the issue?
Please feel free to close the issue if you think so.
Once again, thanks a lot for responding and giving suggestions that helped resolve the issue.

**Comment by s-rog** [comment]
Ah I see, it was a simple typo. Regarding 3, that is intended behavior, glad it's working now and no problem!
